code42/py42
# Refactor, .netrc support, and socket leak fix

### Description of Change ###

- Adds support for .netrc
- Fixes socket leaking issue

Substantially reorganizes the codebase, namely:
- the notion of `_internal` modules has been removed. This was confusing; the only "internal" objects should be those that have nothing to do with business logic, i.e. session / authentication management modules. Those modules are now prefixed with a `_`.
- "Module" classes are now called Clients. The term "Module" is overloaded in python, and "client" seemed appropriate since it's the user will be most directly interfacing with to make requests.
- What we used to have called Clients are now called Services. This again seemed more fitting since they can be mapped one-for-one to corresponding web services.
- Py42Session (and its corresponding variables) has been renamed Connection to better distinguish it from the `requests` object type `Session`.
- MicroserviceClientFactory has been replaced by a `HostResolver` system that allows Connections to be created before their root domain is known (the connection will resolve the root domain when it makes its first request).
- auth now uses the `requests.AuthBase` classes to handle authentication instead of manually modifying headers.

Most importantly, the following have not changed their public signatures at all:
- everything under py42.sdk
- py42.utils
- py42.response
- py42.exceptions
- py42.usercontext

### Issues Resolved ###
- closes https://jira.corp.code42.com/browse/INTEG-983
- closes https://jira.corp.code42.com/browse/INTEG-1092

### Testing Procedure ###
To test .netrc:
- create a .netrc file with the proper configuration in your home directory
- call `sdk.from_local_account` passing `None` for username and password. You should still be able to do everything!

This is what I have tested in the refactor (thought that certainly doesn't mean that you shouldn't also test):
- stream by md5/sha256
- web restore
- FED security events
- at least one call to each of the clients / microservices
- cli unit tests all pass when this installed into it

### PR Checklist ###
Did you remember to do the below?

- [NA] Add docstrings for any new public parameters / methods / classes
